python/ribasim/ribasim/model.py

```diff
@@ -280,6 +285,107 @@ def write(self, filepath: str | PathLike[str]) -> Path:
         context_file_writing.set({})
         return fn
 
+    def _validate_model(self):
+        df_edge = self.edge.df
+        df_chunks = [node.node.df.set_crs(self.crs) for node in self._nodes()]
+        df_node = pd.concat(df_chunks)
+
+        df_graph = df_edge
+        # Join df_edge with df_node to get to_node_type
+        df_graph = df_graph.join(
+            df_node[["node_type"]], on="from_node_id", how="left", rsuffix="_from"
+        )
+        df_graph = df_graph.rename(columns={"node_type": "from_node_type"})
+
+        df_graph = df_graph.join(
+            df_node[["node_type"]], on="to_node_id", how="left", rsuffix="_to"
+        )
+        df_graph = df_graph.rename(columns={"node_type": "to_node_type"})
+
+        if not self._check_neighbors(
```

Comment:
I would raise the ValueError in the function, and only call the function here. Same for the next statement.